TrackController/src/trackctrl/model/TrackController.java

```diff
@@ -92,54 +114,97 @@ public void setZone(HashMap<Integer, Block> blocks) {
 
       //Create string list upon initialization
       for (Integer i : blocks.keySet()) {
-        blockList.add(i.toString());
+        if(myLine.getBlock(i).isSwitch()) {
+          blockList.add("Switch " + i.toString());
+        } else {
+          blockList.add("Block " + i.toString());
+        }
       }
     }
   }
 
   @Override
   public ArrayList<String> getZone() {
+
+    for (Map.Entry<Integer, Block> b : myZone.entrySet()) {
+      if(b.getValue().isSwitch()) {
+        blockList.add("Switch" + " " + b.getKey().toString());
+      } else {
+        blockList.add("Block" + " " + b.getKey().toString());
+      }
+    }
     return blockList;
   }
 
-  //TODO
+  public Block getBlock(int id) {
+    return myZone.get(id);
+  }
+
+  public String getLine() {
+    return myLine.getLine();
+  }
+
+  @Override
+  public int getBlockCount() {
+    return blockList.size();
+  }
+
+  @Override
+  public boolean getOccupancy(int id) {
+    return myLine.getBlock(id).isOccupied();
+  }
+
   @Override
   public boolean addBlock(Block newBlock) {
-    return false;
+
+    myZone.put(newBlock.getNumber(), newBlock);
+    return myZone.containsValue(newBlock);
   }
 
   @Override
   public boolean hasBlock(int id) {
     return myZone.containsKey(id);
   }
 
-  //TODO
   @Override
   public boolean closeBlock(int id) {
-    return false;
+    myLine.getBlock(id).setClosedForMaintenance(true);
+    return myLine.getBlock(id).getBrokenRailStatus() == true;
   }
 
-  //TODO
   @Override
   public boolean repairBlock(int id) {
-    return false;
+    myLine.getBlock(id).setClosedForMaintenance(false);
+    return myLine.getBlock(id).getBrokenRailStatus() == false;
   }
 
-  //TODO
   @Override
   public boolean toggleSwitch(int id) {
+    Block toggle = myLine.getBlock(id);
+    if (toggle.isSwitch()) {
+      Switch toggleSwitch = (Switch) toggle;
+      toggleSwitch.toggle();
+      return toggleSwitch.getSwitchState();
+    }
     return false;
   }
 
-  //TODO
   @Override
   public boolean toggleCrossing(int id) {
+    Block toggle = myLine.getBlock(id);
+    if (toggle.isCrossing()) {
+      boolean before = toggle.getCrossingStatus();
+      toggle.setCrossing(!before);
+      boolean after = toggle.getCrossingStatus();
+      return before ^ after;
+    }
```

Comment:
Just return the new switch state. Dont need any information about what it was before